src/Build/Construction/Solution/SolutionFile.cs

```diff
@@ -1211,6 +1218,50 @@ internal void ParseNestedProjects()
 
                 proj.ParentProjectGuid = parentProjectGuid;
             } while (true);
+
+            ProjectFileErrorUtilities.VerifyThrowInvalidProjectFile(ContainsNoCyclicalNesting(), "SubCategoryForSolutionParsingErrors",
+                new BuildEventFileInfo(FullPath, _currentLineNumber, 0), "SolutionContainsCyclicallyNestedProjectFiles", _solutionFile);
+        }
+
+        private bool ContainsNoCyclicalNesting()
+        {
+            SearchState[] parents = new SearchState[_projectsInOrder.Count];
+            for (int a = 0; a < _projectsInOrder.Count; a++) {
+                parents[a] = SearchState.Unchecked;
+            }
+            for (int i = 0; i < _projectsInOrder.Count; i++) {
+                if (!AncestorsContainNoCycle(parents, i))
+                {
+                    return false;
+                }
+            }
+            return true;
+        }
+
+        private bool AncestorsContainNoCycle(SearchState[] parents, int index)
+        {
+            string parentGuid = _projectsInOrder[index].ParentProjectGuid;
+            if (parentGuid == null)
+            {
+                parents[index] = SearchState.AncestorsContainNoCycles;
+                return true;
+            }
+            int parent = _projectsInOrder.IndexOf(_projects[parentGuid]);
```

Comment:
A project can have multiple parents right? What if the cycle is not on the first parent that's found?  `a->[b, c], b->d, c->e, e->[f, a], d->f`
Add a test case for this.

Reply:
Not in this case; it's what you're directly nested under.